enhancements/windows-containers/windows-machine-config-operator.md

```diff
@@ -0,0 +1,186 @@
+* Provide workflows for adding and removing Windows instances provisioned using
+  the Machine API to OpenShift clusters
+* Perform all the required steps within the VM for it to be added to the cluster
+  as an OpenShift worker node and removed from it.
+
+### Non-Goals
+
+As part of this enhancement we do not plan to support:
+* Installing the container runtime in the Windows node
+* Upgrades (This will be addressed in a separate enhancement)
+  * Windows operating system upgrades
+  * Node management like k8s compute component upgrades, reboots and node
+    draining
+* OpenShift Builds
+* OpenShift DeploymentConfigs
+
+## Proposal
+
+The two main pieces that will allow us to achieve our goals is the Windows
+Machine Config operator (WMCO) and a Windows Machine Config Bootstrapper (WMCB)
+binary. WMCO will collect information from the cluster and transfer it to the
+Windows node. The WMCB will then perform the necessary steps on the node for it
+to join the cluster. [WMCB](https://github.com/openshift/windows-machine-config-bootstrapper)
+has already been developed as part of the Windows Containers
+[dev preview](https://github.com/openshift/windows-machine-config-bootstrapper/blob/master/tools/ansible/docs/ocp-4-4-with-windows-server.md).
+
+### Justification
+
+The reason we are using a WMCB binary rather than a container image is
+that all Windows container images are packaged with a Windows kernel and Red Hat
+has a policy to not ship 3rd party kernels for support reasons.
+
+### Design Details
+
+The WMCO image will be packaged with the following binaries:
+* Kubelet
+* CNI plugins
+* Hybrid-overlay
+* WMCB
+* Kube-proxy
+
+WMCO will be published to OperatorHub as a Red Hat operator. The way the cluster
+admin will enable Windows workloads on their cluster would be to first install
+WMCO. WMCO expects the cluster admin to create a predetermined Secret in its
+namespace containing a private key that will be used to interact with the
+Windows instance. WMCO will check for this Secret during boot up time and create
+a user data Secret which the cluster admin can reference in the Windows
+MachineSet that they create as described below. WMCO will populate the user data
+secret with a public key that corresponds to the private key. In addition it
+will have all the pre-requisite steps needed for setting the VM up for key based
+SSH connections.
+
+When the cluster admin wants to bring up Windows worker nodes, they will create
+a [MachineSet](https://github.com/openshift/machine-api-operator/blob/master/config/machineset.yaml)
+where they reference a Windows OS image which has the Docker container runtime
+add-on enabled. They will also specify the user data secret created by WMCO as
+described above. WMCO will watch for Machine objects with a special label that
+indicate that they map to Windows VMs. Once the Machine object is in the
+"provisioned" state, WMCO will connect to the associated VM to transfer the
```

Comment:
s/state/phase. Can we include a link here https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/machine-instance-lifecycle.md

Reply:
Sure.

Reply:
Addressed.